src/java_bytecode/mz_zip_archive.cpp

```diff
@@ -0,0 +1,76 @@
+/*******************************************************************\
+
+Module: mz_zip library wrapper
+
+Author: DiffBlue Ltd
+
+\*******************************************************************/
+
+#include "mz_zip_archive.h"
+#include <stdexcept>
+#include <string>
+#include <vector>
+#define _LARGEFILE64_SOURCE 1
+#include <miniz/miniz.h>
+
+// Original struct is an anonymous struct with a typedef, This is
+// required to remove internals from the header file
+struct mz_zip_archive_statet:public mz_zip_archive { };
+
+mz_zip_archivet::mz_zip_archivet(const std::string &filename):
+  m_state(new mz_zip_archive_statet { })
+{
+  static_assert(sizeof(mz_uint)<=sizeof(size_t),
+                "size_t cannot store mz_zip file ids, choose a larger type");
+  if(MZ_TRUE!=mz_zip_reader_init_file(m_state.get(), filename.data(), 0))
+    throw std::runtime_error("MZT: Could not load a file: "+filename);
+}
+
+mz_zip_archivet::mz_zip_archivet(mz_zip_archivet &&other):
+  m_state(std::move(other.m_state)) { }
+
+mz_zip_archivet::~mz_zip_archivet()
+{
+  if(m_state)
+    mz_zip_reader_end(m_state.get());
```

Comment:
Seeing as you only run this if `m_state` is valid, it seems like it should be in a custom deleter for the unique_ptr, rather than in this class' destructor.